datafusion/expr/src/logical_plan/builder.rs

```diff
@@ -458,6 +461,81 @@ impl LogicalPlanBuilder {
         self.join_detailed(right, join_type, join_keys, filter, false)
     }
 
+    /// Apply a join with on constraint which the key is an expression.
+    ///
+    /// Filter expression expected to contain non-equality predicates that can not be pushed
+    /// down to any of join inputs.
+    /// In case of outer join, filter applied to only matched rows.
+    pub fn join_with_expr_keys(
+        &self,
+        right: &LogicalPlan,
+        join_type: JoinType,
+        join_keys: (Vec<impl Into<Expr>>, Vec<impl Into<Expr>>),
```

Comment:
Given all the logic I think we now have in the optimizer (e.g. eliminate cross join https://github.com/apache/arrow-datafusion/pull/4185) I recommend we *ONLY* support specifying `filter` on this function and leave it up to the optimizer pass to sort out which are eqijoins and which are not

Reply:
I think so. -- We should "believe" our optimizer, If it doesn't perform well, we can polish it!

Reply:
We should definitely test it, but I think the other passes will handle pushing predicates into Joins correctly -- for example https://github.com/apache/arrow-datafusion/blob/209c26670da4ee52cf13e4302250069a686fdaa2/datafusion/optimizer/src/filter_push_down.rs#L406-L443

Reply:
Thank you, got it. I will try it.